### PR TITLE
Formal Documentation

### PR DESCRIPTION
**Pull request by Tokazama**
I figured anything that gets the discussion started right now is better than nothing, so consider this the super-beta-alpha-draft. In terms of basic structure, I figured we could have a basic introduction, a bunch of simple examples, and an API with docs for methods that are ready for public use and experimental/private use. I'm assuming we'll iron out the details of what constitutes and official part of the interface throughout this process.

### CODE REVIEW

File: docs/src/examples.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Examples
+
+## Simple Array Wrapper
+
+```julia
+struct Wrapper{T,N,P<:AbstractArray{T,N}} <: AbstractArray{T,N}
+    parent::P
+end
+
+ArrayInterface.parent_type(:Type{<:Wrapper{T,N,P}}) where {T,N,P} = P
```
Comment (chriselrod) [inline docs/src/examples.md:10]: Missing a `:` in front of `Type`.

File: docs/src/examples.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+Base.parent(x::Wrapper) = x.parent
+
+function ArrayInterface.unsafe_set_element!(x::Wrapper, val, inds)
+    return ArrayInterface.unsafe_set_element!(parent(x), val inds)
```
Comment (chriselrod) [inline docs/src/examples.md:15]: comma between `val` and `inds`

File: docs/src/examples.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+end
+
+function ArrayInterface.unsafe_get_element(x::Wrapper, inds)
+    return ArrayInterface.unsafe_get_element(parent(x), val inds)
```
Comment (chriselrod) [inline docs/src/examples.md:19]: Same here (comma)

**Comment by chriselrod** [comment]
The thing I'd like to both documentation for is the minimal interface needed to add LoopVectorization support.
It requires a fair number of methods, which is why a simple "inherits form parent" trait for wrappers may be nice.
```julia
using ArrayInterface
struct Wrapper{T,N,P<:AbstractArray{T,N}} <: AbstractArray{T,N}
    parent::P
end
ArrayInterface.parent_type(::Type{<:Wrapper{T,N,P}}) where {T,N,P} = P
Base.parent(x::Wrapper) = x.parent
function ArrayInterface.unsafe_set_element!(x::Wrapper, val, inds)
    return ArrayInterface.unsafe_set_element!(parent(x), val, inds)
end
function ArrayInterface.unsafe_get_element(x::Wrapper, inds)
    return ArrayInterface.unsafe_get_element(parent(x), val, inds)
end
function Base.getindex(x::Wrapper, args...; kwargs...)
    return ArrayInterface.getindex(x, args...; kwargs...)
end
function Base.setindex!(x::Wrapper, val, args...; kwargs...)
    return ArrayInterface.setindex!(x, val, args...; kwargs...)
end

A = rand(3,4);
W = Wrapper(A);
```
Now:
```julia
julia> using LoopVectorization
[ Info: Precompiling LoopVectorization [bdcacae8-1622-11e9-2a5c-532679323890]

julia> LoopVectorization.check_args(A)
true

julia> LoopVectorization.check_args(W)
true
```
`LoopVectorization.check_args` passed because
```julia
julia> ArrayInterface.device(A) === ArrayInterface.CPUPointer()
true
```
However, `pointer` wasn't actually defined:
```julia
julia> stridedpointer(A)
VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{StaticInt{8}, Int64}, Tuple{StaticInt{1}, StaticInt{1}}}(Ptr{Float64} @0x00007f99519e1700, (static(8), 24), (static(1), static(1)))

julia> stridedpointer(W)
ERROR: conversion to pointer not defined for Wrapper{Float64, 2, Matrix{Float64}}
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] unsafe_convert(#unused#::Type{Ptr{Float64}}, a::Wrapper{Float64, 2, Matrix{Float64}})
   @ Base ./pointer.jl:67
 [3] pointer(x::Wrapper{Float64, 2, Matrix{Float64}})
   @ Base ./abstractarray.jl:1110
 [4] memory_reference
   @ ~/.julia/dev/VectorizationBase/src/strided_pointers/stridedpointers.jl:36 [inlined]
 [5] memory_reference
   @ ~/.julia/dev/VectorizationBase/src/strided_pointers/stridedpointers.jl:35 [inlined]
 [6] stridedpointer(A::Wrapper{Float64, 2, Matrix{Float64}})
   @ VectorizationBase ~/.julia/dev/VectorizationBase/src/strided_pointers/stridedpointers.jl:71
 [7] top-level scope
   @ REPL[39]:1
```
We could read the error message, and figure out we have to add
```julia
julia> Base.unsafe_convert(::Type{Ptr{T}}, W::Wrapper) where {T} = Base.unsafe_convert(Ptr{T}, parent(W))

julia> stridedpointer(W)
ERROR: MethodError: no method matching iterate(::Nothing)
Closest candidates are:
  iterate(::Union{LinRange, StepRangeLen}) at range.jl:744
  iterate(::Union{LinRange, StepRangeLen}, ::Int64) at range.jl:744
  iterate(::T) where T<:Union{Base.KeySet{var"#s79", var"#s78"} where var"#s78"<:Dict where var"#s79", Base.ValueIterator{var"#s77"} where var"#s77"<:Dict} at dict.jl:693
  ...
Stacktrace:
 [1] iterate
   @ ./generator.jl:44 [inlined]
 [2] collect(itr::Base.Generator{Nothing, Type{Int64}})
   @ Base ./array.jl:678
 [3] map(f::Type, A::Nothing)
   @ Base ./abstractarray.jl:2317
 [4] asvalint(r::Nothing)
   @ VectorizationBase ~/.julia/dev/VectorizationBase/src/VectorizationBase.jl:15
 [5] val_stride_rank
   @ ~/.julia/dev/VectorizationBase/src/VectorizationBase.jl:16 [inlined]
 [6] stridedpointer(A::Wrapper{Float64, 2, Matrix{Float64}})
   @ VectorizationBase ~/.julia/dev/VectorizationBase/src/strided_pointers/stridedpointers.jl:71
 [7] top-level scope
   @ REPL[45]:1
```
But now it'd probably be hard to connect the dots between `val_stride_rank` and `ArrayInterface.stride_rank`.
I realize I should improve this with better error messages inside VectorizationBase, e.g. if `stride_rank` returns `nothing`, `val_stride_rank` should instruct the users on what to do, and also provide a helpful link to this documentation once its merged.

The list of additional methods:
```julia
Base.size(W::Wrapper) = Base.size(parent(W))
for f in [:strides, :stride_rank, :contiguous_axis, :contiguous_batch_size, :dense_dims] 
    @eval ArrayInterface.$f(W::Wrapper) = ArrayInterface.$f(parent(W))
end
```
`offsets` may be a good idea to add as well. It works without a definition by assuming `One()`.
With `ArrayInterface.size` instead of `Base.size`, I got:
```julia
julia> stridedpointer(W)
ERROR: MethodError: no method matching size(::Wrapper{Float64, 2, Matrix{Float64}})
Closest candidates are:
  size(::AbstractArray{T, N}, ::Any) where {T, N} at abstractarray.jl:38
  size(::Union{LinearAlgebra.Adjoint{T, var"#s830"}, LinearAlgebra.Transpose{T, var"#s830"}} where var"#s830"<:(AbstractVector{T} where T) where T) at /home/chriselrod/Documents/languages/julia/usr/share/julia/stdlib/v1.7/LinearAlgebra/src/adjtrans.jl:208
  size(::Union{LinearAlgebra.Adjoint{T, var"#s830"}, LinearAlgebra.Transpose{T, var"#s830"}} where var"#s830"<:(AbstractMatrix{T} where T) where T) at /home/chriselrod/Documents/languages/julia/usr/share/julia/stdlib/v1.7/LinearAlgebra/src/adjtrans.jl:209
  ...
Stacktrace:
  [1] axes
    @ ./abstractarray.jl:89 [inlined]
  [2] axes
    @ ~/.julia/dev/ArrayInterface/src/dimensions.jl:359 [inlined]
  [3] axes
    @ ~/.julia/dev/ArrayInterface/src/dimensions.jl:352 [inlined]
  [4] indices
    @ ~/.julia/dev/ArrayInterface/src/ranges.jl:581 [inlined]
  [5] offsets
    @ ~/.julia/dev/ArrayInterface/src/stridelayout.jl:9 [inlined]
  [6] macro expansion
    @ ~/.julia/dev/ArrayInterface/src/static.jl:289 [inlined]
  [7] eachop
    @ ~/.julia/dev/ArrayInterface/src/static.jl:289 [inlined]
  [8] offsets(x::Wrapper{Float64, 2, Matrix{Float64}})
    @ ArrayInterface ~/.julia/dev/ArrayInterface/src/stridelayout.jl:11
  [9] stridedpointer(A::Wrapper{Float64, 2, Matrix{Float64}})
    @ VectorizationBase ~/.julia/dev/VectorizationBase/src/strided_pointers/stridedpointers.jl:71
 [10] top-level scope
    @ REPL[53]:1
```
Probably better to fallback on `Base.axes`
```julia
"""
    axes(A)

Return a tuple of ranges where each range maps to each element along a dimension of `A`.
"""
@inline axes(A) = Base.axes(A)
```
instead of trying `ArrayInterface.size` and `ArrayInterface.offsets`?

All that aside, now:
```julia
julia> stridedpointer(W)
VectorizationBase.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{StaticInt{8}, Int64}, Tuple{StaticInt{1}, StaticInt{1}}}(Ptr{Float64} @0x00007f99519e1700, (static(8), 24), (static(1), static(1)))
```
And LoopVectorization should work as expected.

The `@eval` loop above would be pretty easy to copy, paste, and edit if something like it is in the example, so perhaps it's not really necessary to provide some automation on our own.
They're also very likely to be defining `Base.size` anyway, and by adding the module to the function names, we could stick it inside the loop too:
```julia
for f in [:(Base.size), :(ArrayInterface.strides), :(ArrayInterface.stride_rank), :(ArrayInterface.contiguous_axis), :(ArrayInterface.contiguous_batch_size), :(ArrayInterface.dense_dims)] 
    @eval $f(W::Wrapper) = $f(parent(W))
end
```
So that's not much boiler plate.

**Comment by Tokazama** [comment]
If we add a fall back for `stride_rank` do you think we could circumvent people adding all the stride related methods? I think it would be better if most people don't mess with strides at all.

**Comment by chriselrod** [comment]
> If we add a fall back for `stride_rank` do you think we could circumvent people adding all the stride related methods? I think it would be better if most people don't mess with strides at all.

I don't think we should discourage it, but I do think it'd be great if this were automated so that it works in all the most typical cases. 

I think we should just assume that `contiguous_batch(A) = Zero()` and `contiguous_axis(A)` is whatever axis has a stride rank of 1 (if none do, then `StaticInt{-1}()`).
It would be rare for this not to be the case, so it should only be necessary to specify when it differs.

`stride_rank` + `dense_dims` + `size` also gives us all the strides as a function of `size`, in case they'd rather specify `dense_dims` than `size`. In that case, we should maybe have a generic `strides` that uses `dense_dims` before falling back on `Base.strides` (and also inserting `One()` as appropriate for the `contiguous_axis`, as it does currently).

**Comment by Tokazama** [comment]
> I don't think we should discourage it, but I do think it'd be great if this were automated so that it works in all the most typical cases.

I just mean that the vast majority of people aren't working with something that has linear indexing and changes strides. I think it would make sense to cover when users need to define these in the examples though.

I'll try to put together a minimal wrapper in the tests. There may be a couple minor changes to internal code to make it work, but I'll steer away from anything breaking or controversial.

**Comment by chriselrod** [comment]
> I just mean that the vast majority of people aren't working with something that has linear indexing and changes strides. I think it would make sense to cover when users need to define these in the examples though.

I agree. I think largely that if someone is doing something out of the norm enough that the defaults aren't incorrect, they have a reason for what they're doing and will be interested in defining these methods to specify the exact behavior they want anyway.

**Comment by Tokazama** [comment]
The latest commit won't work but I wanted a good checkpoint explaining most the changes b/c I'll likely be busy for the next couple days. I think the only thing keeping it from working is getting `strides` working for `ReinterpretArray`. I shamelessly copied some code from Base to get it working but I need to look at it more.

I'd like to ultimately get us away from relying on the base version of methods we implement, b/c I think that's the only way to robustly stress test this interface to all the demands of a general array interface (e.g., I threw in a bunch of `axes(x::BaseArrayType) = Base.axes(x)` that we could probably do a better job with in the future).

I probably should also do some basic checks for inheriting traits through a parents. For example, if the parent type has a different number of dimensions it can't have the same strides.

**Comment by chriselrod** [comment]
Strides are still broken:
```julia
julia> using ArrayInterface

julia> colors = [(R = rand(), G = rand(), B = rand()) for i ∈ 1:100];

julia> colormat = reinterpret(reshape, Float64, colors)
3×100 reinterpret(reshape, Float64, ::Vector{NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}}) with eltype Float64:
 0.128668  0.471231  0.997608  0.411756  0.634709  0.737467  0.666604  0.789188  0.20917   0.221311  0.747995   0.654684  0.779321  0.00807382  0.638984  0.906769  0.339308  …  0.566442  0.744517  0.542496   0.617728  0.68426   0.00993767  0.39687   0.954698  0.581257  0.585789  0.931305  0.894563  0.358154  0.740562  0.620136  0.127963
 0.887468  0.564366  0.980133  0.114801  0.60867   0.373185  0.524908  0.494138  0.161197  0.247497  0.0925153  0.344362  0.216621  0.0510951   0.412955  0.499191  0.572357     0.9878    0.127503  0.0806957  0.650491  0.648738  0.362777    0.625217  0.505443  0.351756  0.106991  0.417529  0.944718  0.451381  0.421518  0.536598  0.51561
 0.391843  0.140135  0.515798  0.763666  0.854456  0.368541  0.282564  0.602923  0.662462  0.955656  0.722027   0.32824   0.449584  0.383805    0.892799  0.72917   0.698284     0.988841  0.42103   0.775026   0.15183   0.869753  0.315611    0.217497  0.172563  0.529196  0.862757  0.946572  0.422325  0.436635  0.657204  0.869399  0.163631

julia> ArrayInterface.strides(colormat)
ERROR: UndefVarError: S not defined
Stacktrace:
 [1] _reinterpret_size(x::Base.ReinterpretArray{Float64, 2, NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}, Vector{NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}}, true}, #unused#::True)
   @ ArrayInterface ~/.julia/dev/ArrayInterface/src/reinterpretarray.jl:31
 [2] size
   @ ~/.julia/dev/ArrayInterface/src/reinterpretarray.jl:26 [inlined]
 [3] strides(a::Base.ReinterpretArray{Float64, 2, NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}, Vector{NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}}, true})
   @ ArrayInterface ~/.julia/dev/ArrayInterface/src/reinterpretarray.jl:75
 [4] top-level scope
   @ REPL[5]:1
```
Versus master:
```julia
julia> using ArrayInterface
[ Info: Precompiling ArrayInterface [4fba245c-0d91-5ea0-9b3e-6abc04ee57a9]

julia> colors = [(R = rand(), G = rand(), B = rand()) for i ∈ 1:100];

julia> colormat = reinterpret(reshape, Float64, colors)
3×100 reinterpret(reshape, Float64, ::Vector{NamedTuple{(:R, :G, :B), Tuple{Float64, Float64, Float64}}}) with eltype Float64:
 0.409459  0.76078   0.328358  0.993605  0.733902  0.573292  0.913654  0.111111  0.238478  0.305678  0.634185  0.474198  0.753014  0.873242  0.712745  0.627483  0.720143  …  0.448289  0.44624   0.427579   0.455781  0.761577  0.0748551  0.234329   0.443285  0.530865    0.655775   0.354764   0.85953   0.955834  0.966037  0.366734  0.22914
 0.201132  0.582454  0.771959  0.25598   0.372418  0.396411  0.98311   0.959752  0.700328  0.405852  0.610949  0.865397  0.416433  0.930137  0.884745  0.279771  0.159928     0.863986  0.734954  0.0514181  0.583819  0.456161  0.894806   0.28982    0.722886  0.00701849  0.369287   0.0809389  0.818646  0.273236  0.394714  0.340524  0.895546
 0.650679  0.456191  0.388047  0.567046  0.760825  0.321318  0.672562  0.236187  0.451545  0.778551  0.665279  0.138917  0.271032  0.465364  0.889064  0.91338   0.140144     0.175752  0.670934  0.763592   0.444447  0.744188  0.242061   0.0215222  0.558682  0.987231    0.0505401  0.975547   0.997723  0.487984  0.306042  0.833296  0.52108

julia> ArrayInterface.strides(colormat)
(static(1), static(3))
```
Getting static strides is important here.

**Comment by Tokazama** [comment]
Sounds good. I can put everything back where it belongs as this gets cleaned up (e.g., get rid of the "reinterpretarray.jl" file).

**Comment by chriselrod** [comment]
>  If I understand correctly, this is more like an extension of the Base AbstractArray interface, where what the OP seems to be looking for is something like Tables.jl for ND arrays with labels for any/all axes. Is that right?
> 
> Or is the point that, if all of the current AxisArray-like types implemented this interface, we’d get that interoperability?

https://discourse.julialang.org/t/converting-between-axisarray-like-packages/52510/6?u=elrod

Could do with an example using `has_dimnames` and getting kwarg indexing to work.

**Comment by Tokazama** [comment]
Will document dimensions specifically in another PR.

**Pull request closed without merging**